File: customize_erpnext/customize_erpnext/report/bom_balance_sheet/bom_balance_sheet.py
# Copyright (c) 2025, Your Company and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.utils import flt
from erpnext.stock.utils import get_stock_balance


def execute(filters=None):
    if not filters:
        filters = {}
        
    # Nếu có sale order được chọn, lấy danh sách item templates liên quan
    if filters.get("sale_order"):
        item_templates = get_item_templates_from_sales_order(filters.get("sale_order"))
        filters["available_templates"] = item_templates
        
        # Extract item template name từ "name:label" format nếu cần
        item_template = extract_item_template(filters.get("item_template"))
        if item_template:
            filters["item_template"] = item_template
        
        # Nếu có item_template được chọn nhưng không nằm trong danh sách liên quan, thông báo lỗi
        if item_template and item_template not in item_templates:
            frappe.throw(_("Selected Item Template is not related to the chosen Sales Order"))

    columns = get_columns(filters)
    data = get_data(filters)

    return columns, data


def get_columns(filters=None):
    columns = [
        {
            "fieldname": "item_group",
            "fieldtype": "Link",
            "label": _("Item Group"),
            "options": "Item Group",
            "width": 120
        },
        {
            "fieldname": "item",
            "fieldtype": "Link",
            "label": _("Item"),
            "options": "Item",
            "width": 150
        },
        {
            "fieldname": "description",
            "fieldtype": "Data",
            "label": _("Description"),
            "width": 200
        },
        {
            "fieldname": "item_name",
            "fieldtype": "Data",
            "label": _("Item Name"),
            "width": 150
        },
        {
            "fieldname": "color",
            "fieldtype": "Data",
            "label": _("Color"),
            "width": 150
        },
        {
            "fieldname": "size",
            "fieldtype": "Data",
            "label": _("Size"),
            "width": 100
        },
        {
            "fieldname": "lost_percent",
            "fieldtype": "Float",
            "label": _("Lost Percent"),
            "width": 100,
            "precision": 2
        },
        {
            "fieldname": "quantity_require",
            "fieldtype": "Float",
            "label": _("Quantity Require"),
            "width": 120,
            "precision": 2
        },
        {
            "fieldname": "quantity_require_include_lost_percent",
            "fieldtype": "Float",
            "label": _("Qty Require Include Lost Percent"),
            "width": 200,
            "precision": 2
        },  
        {
            "fieldname": "quantity_available_in_stock",
            "fieldtype": "Float",
            "label": _("Qty Available in Stock"),
            "width": 150,
            "precision": 2
    }
    ]
    
 
    
    return columns


def get_data(filters):
    data = []
    
    # If no sale order and no item template, return empty data
    if not filters.get("sale_order") and not filters.get("item_template"):
        return data
        
    boms = get_boms_from_filters(filters)
    
    if not boms:
        return data
        
    items_data = {}
    
    # Kiểm tra xem custom field đã tồn tại chưa
    has_custom_lost = frappe.db.exists(
        "Custom Field",
        {"dt": "BOM Item", "fieldname": "custom_lost"}
    )
    
    # Kiểm tra có sử dụng with_percent_lost không
    with_percent_lost = filters.get("with_percent_lost", 1)
    
    for bom in boms:
        bom_items = get_bom_items(bom.name)
        parent_qty = get_parent_qty_from_filters(filters, bom.item)
        
        for item in bom_items:
            total_qty = flt(parent_qty) * flt(item.qty)
            
            # Lấy lost_percent từ custom field nếu có, nếu không thì mặc định là 0
            lost_percent = 0
            if has_custom_lost and hasattr(item, 'custom_lost'):
                lost_percent = flt(item.custom_lost) or 0
            
            # Tính toán quantity với lost percent
            if with_percent_lost:
                qty_with_lost = total_qty + (total_qty * lost_percent / 100)
            else:
                qty_with_lost = total_qty
                
            key = item.item_code
            
            if key in items_data:
                items_data[key]["quantity_require"] += total_qty
                if with_percent_lost:
                    items_data[key]["quantity_require_include_lost_percent"] += qty_with_lost
            else:
                # Lấy thông tin color, size và item_group
                color, size = get_item_attributes(item.item_code)
                item_group = get_item_group(item.item_code)
                
                default_warehouse = get_default_warehouse()
                stock_qty = 0
                if default_warehouse:
                    try:
                        stock_qty = get_stock_balance(item.item_code, default_warehouse)
                    except:
                        stock_qty = 0
                
                item_data = {
                    "item_group": item_group,
                    "item": item.item_code,
                    "description": item.description,
                    "item_name": item.item_name,
                    "color": color,
                    "size": size,
                    "lost_percent": lost_percent,
                    "quantity_require": total_qty,
                    "quantity_available_in_stock": stock_qty,
                    "with_percent_lost": with_percent_lost
                }
                
                # Only add lost percent quantity if the option is enabled
                if with_percent_lost:
                    item_data["quantity_require_include_lost_percent"] = qty_with_lost
                    
                items_data[key] = item_data
    
    # Convert the dictionary to a list for the report
    data = list(items_data.values())
    
    # Sort data by item code only (no grouping)
    data.sort(key=lambda x: x["item"])
    
    return data


def get_boms_from_filters(filters):
    if filters.get("sale_order") and not filters.get("item_template"):
        # Nếu chỉ có sale order thì lấy tất cả BOMs từ sale order (hiển thị toàn bộ nguyên phụ liệu)
        return get_boms_from_sales_order(filters)
    elif filters.get("sale_order") and filters.get("item_template"):
        # Nếu có cả sale order và item template
        if filters.get("color"):
            return get_boms_from_template_and_color(filters)
        else:
            return get_boms_from_template(filters)
    elif filters.get("item_template"):
        # Nếu chỉ có item template
        if filters.get("color"):
            return get_boms_from_template_and_color(filters)
        else:
            return get_boms_from_template(filters)
    
    return []


def get_boms_from_sales_order(filters):
    sales_order_items = frappe.get_all(
        "Sales Order Item",
        filters={"parent": filters.get("sale_order")},
        fields=["item_code", "qty"]
    )
    
    boms = []
    for item in sales_order_items:
        bom = frappe.get_all(
            "BOM",
            filters={
                "item": item.item_code,
                "is_active": 1,
                "is_default": 1
            },
            fields=["name", "item"]
        )
        
        if bom:
            boms.extend(bom)
    
    return boms


def get_boms_from_template_and_color(filters):
    # Extract item template name từ "name:label" format nếu cần
    item_template = extract_item_template(filters.get("item_template"))
    
    # If filters has available_templates and the current template is not in it, return empty
    if filters.get("available_templates") and item_template not in filters.get("available_templates"):
        return []
        
    # Lấy các variants của item template có color được chọn
    variant_items = frappe.db.sql("""
        SELECT i.name
        FROM `tabItem` i
        JOIN `tabItem Variant Attribute` iva ON i.name = iva.parent
        WHERE i.variant_of = %s
        AND iva.attribute = 'Color'
        AND iva.attribute_value = %s
    """, (item_template, filters.get("color")), as_dict=1)
    
    boms = []
    for item in variant_items:
        bom = frappe.get_all(
            "BOM",
            filters={
                "item": item.name,
                "is_active": 1,
                "is_default": 1
            },
            fields=["name", "item"]
        )
        
        if bom:
            boms.extend(bom)
    
    return boms


def get_boms_from_template(filters):
    # Extract item template name từ "name:label" format nếu cần
    item_template = extract_item_template(filters.get("item_template"))
    
    # If filters has available_templates and the current template is not in it, return empty
    if filters.get("available_templates") and item_template not in filters.get("available_templates"):
        return []
        
    variant_items = frappe.db.sql("""
        SELECT i.name
        FROM `tabItem` i
        WHERE i.variant_of = %s
    """, item_template, as_dict=1)
    
    boms = []
    for item in variant_items:
        bom = frappe.get_all(
            "BOM",
            filters={
                "item": item.name,
                "is_active": 1,
                "is_default": 1
            },
            fields=["name", "item"]
        )
        
        if bom:
            boms.extend(bom)
    
    return boms


def get_bom_items(bom_name):
    # Kiểm tra xem custom field đã tồn tại chưa
    has_custom_lost = frappe.db.exists(
        "Custom Field",
        {"dt": "BOM Item", "fieldname": "custom_lost"}
    )
    
    # Chuẩn bị danh sách trường cần lấy
    fields = ["item_code", "item_name", "description", "qty"]
    
    # Chỉ thêm custom_lost vào fields nếu nó tồn tại
    if has_custom_lost:
        fields.append("custom_lost")
    
    bom_items = frappe.get_all(
        "BOM Item",
        filters={"parent": bom_name},
        fields=fields
    )
    
    return bom_items


def get_parent_qty_from_filters(filters, item_code):
    if filters.get("sale_order"):
        sales_order_item = frappe.get_all(
            "Sales Order Item",
            filters={
                "parent": filters.get("sale_order"),
                "item_code": item_code
            },
            fields=["qty"]
        )
        
        if sales_order_item:
            return sales_order_item[0].qty
    
    return 1  # Default quantity if not specified


def get_default_warehouse():
    default_warehouse = frappe.db.get_single_value("Stock Settings", "default_warehouse")
    
    if not default_warehouse:
        # Fallback to a company's default warehouse if Stock Settings doesn't have one
        company = frappe.defaults.get_user_default("Company")
        if company:
            default_warehouse = frappe.db.get_value("Company", company, "default_warehouse")
    
    return default_warehouse


def get_item_attributes(item_code):
    """Get color and size attributes for an item"""
    attributes = frappe.db.sql("""
        SELECT attribute, attribute_value
        FROM `tabItem Variant Attribute`
        WHERE parent = %s AND attribute IN ('Color', 'Size')
    """, item_code, as_dict=1)
    
    color = ""
    size = ""
    
    for attr in attributes:
        if attr.attribute == "Color":
            color = attr.attribute_value
        elif attr.attribute == "Size":
            size = attr.attribute_value
    
    return color, size


def get_item_group(item_code):
    """Get item group for an item"""
    item_group = frappe.db.get_value("Item", item_code, "item_group")
    return item_group or ""

def extract_item_template(item_template_filter):
    """Extract item template name từ 'name:label' format nếu cần"""
    if item_template_filter and ":" in item_template_filter:
        return item_template_filter.split(":")[0]
    return item_template_filter

@frappe.whitelist()
def get_item_templates_from_sales_order(sales_order):
    """Get all B-Finished Goods item templates related to items in a sales order"""
    templates = frappe.db.sql("""
        SELECT DISTINCT i.variant_of as template
        FROM `tabSales Order Item` soi
        JOIN `tabItem` i ON soi.item_code = i.name
        WHERE soi.parent = %s 
        AND i.variant_of IS NOT NULL
        AND i.variant_of != ''
        AND i.is_sales_item = 1
        AND i.disabled = 0
        AND i.variant_of LIKE 'B-%%'
    """, sales_order, as_dict=1)
    
    return [t.template for t in templates if t.template]

@frappe.whitelist()
def get_filtered_item_templates(sales_order=None):
    """
    Get filtered item templates for Select field based on sales order
    Returns data formatted for ERPNext Select field
    """
    try:
        if sales_order:
            # Get item templates from specific sales order
            templates = get_item_templates_from_sales_order(sales_order)
            
            if not templates:
                return []
            
            # Get full item details for these templates
            item_details = frappe.db.sql("""
                SELECT 
                    item.name,
                    item.item_name,
                    item.description
                FROM `tabItem` item
                WHERE 
                    item.name IN %(templates)s
                    AND item.has_variants = 1
                    AND item.is_sales_item = 1
                    AND item.disabled = 0
                    AND item.name LIKE 'B-%%'
                ORDER BY item.item_name
            """, {
                'templates': templates
            }, as_dict=True)
            
            return item_details
            
        else:
            # Return all B-Finished Goods templates when no sales order
            item_details = frappe.db.sql("""
                SELECT 
                    item.name,
                    item.item_name,
                    item.description
                FROM `tabItem` item
                WHERE 
                    item.has_variants = 1
                    AND item.is_sales_item = 1
                    AND item.disabled = 0
                    AND item.name LIKE 'B-%%'
                ORDER BY item.item_name
                LIMIT 100
            """, as_dict=True)
            
            return item_details
            
    except Exception as e:
        frappe.log_error(f"Error in get_filtered_item_templates: {str(e)}")
        return []

@frappe.whitelist()
def get_filtered_item_templates_with_cache(sales_order=None):
    """
    Enhanced version with caching for better performance
    """
    cache_key = f"bom_item_templates_{sales_order or 'all'}"
    
    # Try to get from cache first (cache for 5 minutes)
    cached_result = frappe.cache().get_value(cache_key)
    if cached_result:
        return cached_result
    
    try:
        result = get_filtered_item_templates(sales_order)
        
        # Cache the result for 5 minutes
        frappe.cache().set_value(cache_key, result, expires_in_sec=300)
        
        return result
        
    except Exception as e:
        frappe.log_error(f"Error in get_filtered_item_templates_with_cache: {str(e)}")
        return []

File: customize_erpnext/customize_erpnext/report/bom_balance_sheet/bom_balance_sheet.js
// Copyright (c) 2025, Your Company and contributors
// For license information, please see license.txt

frappe.query_reports["BOM Balance Sheet"] = {
	"filters": [
		{
			"fieldname": "sale_order",
			"label": __("Sale Order"),
			"fieldtype": "Link",
			"options": "Sales Order",
			"get_query": function () {
				return {
					filters: [
						["Sales Order", "docstatus", "=", 1]
					]
				};
			},
			"on_change": function () {
				// When sales order is selected, update item_template options from server
				const report = frappe.query_report;
				const sale_order = report.get_filter_value('sale_order');

				// Clear dependent filters
				report.set_filter_value('item_template', '');
				report.set_filter_value('color', '');

				// Update item_template options based on sales order
				update_item_template_options(report, sale_order);
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "item_template",
			"label": __("Item Template"),
			"fieldtype": "Select",
			"options": "",
			"reqd": 0,
			"on_change": function () {
				// Reset color filter and populate the color dropdown
				const report = frappe.query_report;
				const item_template_raw = report.get_filter_value('item_template');

				if (item_template_raw) {
					// Extract item template name từ "name:label" format nếu cần
					const item_template = item_template_raw.includes(':') ?
						item_template_raw.split(':')[0] : item_template_raw;

					// Reset and update color filter options
					report.set_filter_value('color', '');

					// Use the existing function to get colors for the template
					frappe.call({
						method: "customize_erpnext.api.utilities.get_colors_for_template",
						args: {
							item_template: item_template  // Pass extracted name
						},
						callback: function (r) {
							if (r.message) {
								const color_filter = report.get_filter('color');
								const color_options = [""].concat(r.message);
								color_filter.df.options = color_options;
								color_filter.refresh();
							}
						}
					});
				} else {
					// Clear color filter when no item template is selected
					const color_filter = report.get_filter('color');
					color_filter.df.options = "";
					color_filter.refresh();
				}
				frappe.query_report.refresh();
			}
		},
		{
			"fieldname": "color",
			"label": __("Color"),
			"fieldtype": "Select",
			"options": "",
			"depends_on": "eval:doc.item_template"
		},
		{
			"fieldname": "with_percent_lost",
			"label": __("Compare Qty Including Lost Percent"),
			"fieldtype": "Check",
			"default": 1,
			"on_change": function () {
				// Refresh report when this filter changes to rebuild columns
				frappe.query_report.refresh();
			}
		}
	],

	"formatter": function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);

		// Format stock quantity colors - compare with the right quantity based on with_percent_lost
		if (column.fieldname == "quantity_available_in_stock" && data) {
			const available = parseFloat(data.quantity_available_in_stock) || 0;
			let required = 0;

			// Check if with_percent_lost is enabled based on data or current filter
			const with_percent_lost = frappe.query_report.get_filter_value('with_percent_lost');

			if (with_percent_lost && data.quantity_require_include_lost_percent !== undefined) {
				required = parseFloat(data.quantity_require_include_lost_percent) || 0;
			} else {
				required = parseFloat(data.quantity_require) || 0;
			}

			// Display numeric value with color
			let displayValue = frappe.format(available, { fieldtype: "Float", precision: 3 });

			if (available < required) {
				value = `<span style="color: #d32f2f; font-weight: bold;">${displayValue}</span>`;
			} else {
				value = `<span style="color: #388e3c; font-weight: bold;">${displayValue}</span>`;
			}
		}

		// Format item code with link
		if (column.fieldname == "item" && data) {
			value = `<a href="/app/item/${data.item}" target="_blank" style="color: #1976d2; text-decoration: none;">${data.item}</a>`;
		}

		// Format lost percent
		if (column.fieldname == "lost_percent" && data && parseFloat(value) > 0) {
			value = `<span style="color: #f57c00; font-weight: 500;">${value}%</span>`;
		}

		return value;
	},

	"onload": function (report) {
		// Initialize item_template options on load
		update_item_template_options(report, null);
		// Add refresh button
		report.page.add_inner_button(__("Refresh"), function () {
			frappe.query_report.refresh();
		});
		// Add custom buttons
		report.page.add_inner_button(__("Export to Excel"), function () {
			frappe.query_report.export_report("Excel");
		});

		// report.page.add_inner_button(__("Print"), function () {
		// 	frappe.query_report.print_report();
		// });


	}
};

// Helper function to update item_template options via server-side filtering
function update_item_template_options(report, sale_order) {
	frappe.call({
		method: "customize_erpnext.customize_erpnext.report.bom_balance_sheet.bom_balance_sheet.get_filtered_item_templates",
		args: {
			sales_order: sale_order || null
		},
		callback: function (r) {
			if (r.message) {
				const item_template_filter = report.get_filter('item_template');

				// Create options string for Select field
				let options = "";
				if (r.message.length > 0) {
					// Add empty option first, then all templates với item_name để dễ nhận biết
					options = "\n" + r.message.map(item => {
						// Format as "value:label" for better display
						return `${item.name}:${item.item_name || item.name}`;
					}).join("\n");
				}

				// Update filter options
				item_template_filter.df.options = options;
				item_template_filter.refresh();

				// Show user feedback
				if (sale_order && r.message.length > 0) {
					frappe.show_alert({
						message: __(`Found ${r.message.length} item templates for Sales Order ${sale_order}`),
						indicator: 'blue'
					});
				} else if (sale_order && r.message.length === 0) {
					frappe.show_alert({
						message: __(`No item templates found for Sales Order ${sale_order}`),
						indicator: 'orange'
					});
				}
			} else {
				// Clear options if no data
				const item_template_filter = report.get_filter('item_template');
				item_template_filter.df.options = "";
				item_template_filter.refresh();
			}
		},
		error: function (r) {
			console.error('Error updating item template options:', r);
			frappe.show_alert({
				message: __('Error loading item templates'),
				indicator: 'red'
			});
		}
	});
}